pramen/core/src/main/scala/za/co/absa/pramen/core/sql/SqlGeneratorMicrosoft.scala

```diff
@@ -97,7 +100,119 @@ class SqlGeneratorMicrosoft(sqlConfig: SqlConfig) extends SqlGeneratorBase(sqlCo
     }
   }
 
+  override def getAliasExpression(expression: String, alias: String): String = {
+    s"$expression AS ${escape(alias)}"
+  }
+
+  override def quote(identifier: String): String = {
+    validateIdentifier(identifier)
+    splitComplexIdentifier(identifier).map(quoteSingleIdentifier).mkString(".")
+  }
+
+  override def escape(identifier: String): String = {
+    if (needsEscaping(sqlConfig.identifierQuotingPolicy, identifier)) {
+      quote(identifier)
+    } else {
+      identifier
+    }
+  }
+
   private def getLimit(limit: Option[Int]): String = {
     limit.map(n => s"TOP $n ").getOrElse("")
   }
+
+  private def columnExpr(columns: Seq[String]): String = {
+    if (columns.isEmpty) {
+      "*"
+    } else {
+      columns.map(col => escape(col)).mkString(", ")
+    }
+  }
+
+  private def infoDateColumn: String = {
+    escape(sqlConfig.infoDateColumn)
+  }
+
+  private def quoteSingleIdentifier(identifier: String): String = {
+    val (escapeBegin, escapeEnd) = beginEndEscapeChars
+
+    if (
+      (identifier.startsWith(s"$escapeBegin") && identifier.endsWith(s"$escapeEnd")) ||
+        (identifier.startsWith(s"$escapeChar2") && identifier.endsWith(s"$escapeChar2"))
+    ) {
+      identifier
+    } else {
+      s"$escapeBegin$identifier$escapeEnd"
+    }
+  }
+
+  private[core] def splitComplexIdentifier(identifier: String): Seq[String] = {
+    val trimmedIdentifier = identifier.trim
+
+    if (trimmedIdentifier.isEmpty) {
+      throw new IllegalArgumentException(f"Found an empty table name or column name ('$identifier').")
+    }
+
+    val (escapeBegin1, escapeEnd1) = beginEndEscapeChars
+
+    val output = new ListBuffer[String]
+    val curColumn = new StringBuffer()
+    val len = trimmedIdentifier.length
+    var nestingChar = new ListBuffer[Char]
+    var i = 0
+
+    while (i < len) {
+      val c = trimmedIdentifier(i)
+      val nextChar = if (i == len - 1) ' ' else trimmedIdentifier(i + 1)
+
+      if (nestingChar.isEmpty && c == '.') {
+        output += curColumn.toString
+        curColumn.setLength(0)
+      } else {
+        curColumn.append(c)
+      }
+
+      if (c == escapeChar2) {
+        if (nestingChar.isEmpty) {
+          nestingChar.prepend(escapeChar2)
+          if (curColumn.length() > 1 && i < len - 1 && nextChar != '.')
```

Comment:
This if-clause needs to move above the `        if (nestingChar.isEmpty) {` condition.

The following test-case throws an exception in the SqlGeneratorBase
```
    "throw on invalid open double-quote" in {
      val ex = intercept[IllegalArgumentException] {
        genDate.asInstanceOf[SqlGeneratorBase].splitComplexIdentifier("\"System User.\"Table Name")
      }

      assert(ex.getMessage.contains("Invalid character '\"' in the identifier '\"System User.\"Table Name'"))
    }
```

However, it doesn't throw an exception in SqlGeneratorMicrosoft

Reply:
Good catch!